js/runtime.js

```diff
@@ -821,51 +823,165 @@
+            playAudio: function(audio){
+                audio.play();
+            },
+            playInstrument: function(sound){
+                console.log(song);
+                T("mml", {mml:song}, sound).on("ended", function() {
+                    sound.pause();
+                    this.stop();
+                }).start();
+                song = "o4 l4 V12 ";
+            },
+            mml: function(sound, mml){
+                var gen = T("OscGen", {wave:sound, env:{type:"perc"}, mul:0.25}).play();
+                T("mml", {mml:mml}, gen).on("ended", function() {
+                    gen.pause();
+                    this.stop();
+                }).start();
+            },
+            tempo: function(tempo){
+                song += ("t" + tempo + " ");
             },
             pause: function(sound){
                 sound.pause();
             },
-            playFrom: function(sound, time){
-                sound.playFrom(time);
-            },
-            pan: function(sound, balance){
-                sound.pan = balance;
-            },
-            echo_DelayFeedbackFilter: function(sound, delay, feedback, filter){
-                sound.setEcho(delay, feedback, filter);
-            },
-            stopEcho: function(sound){
-                sound.echo = false;
-            },
-            reverb_DurationDecayReverse: function(sound, duration, decay, reverse){
-                sound.setReverb(duration, decay, reverse);
-            },
-            stopReverb: function(sound){
-                sound.reverb = false;
+            keys: function(wave, vol){
+                var synth = T("OscGen", {wave:wave, mul:vol}).play();
+
+                var keydict = T("ndict.key");
+                var midicps = T("midicps");
+                T("keyboard").on("keydown", function(e) {
+                  var midi = keydict.at(e.keyCode);
+                  if (midi) {
+                    var freq = midicps.at(midi);
+                    synth.noteOnWithFreq(freq, 100);
+                  }
+                }).on("keyup", function(e) {
+                  var midi = keydict.at(e.keyCode);
+                  if (midi) {
+                    synth.noteOff(midi, 100);
+                  }
+                }).start();
+
+                alert("Play notes on the keyboard");
             },
-            effect: function(frequency, attack, decay, wait, echoDelay, echoFeedback, echoFilter, waveform, volume, balance, pitchBend, reverseBend, random, dissonance){
-                return {
-                    play: function(){
-                        soundEffect(
-                            frequency, attack, decay, waveform,
-                            volume, balance, wait,
-                            pitchBend, reverseBend, random, dissonance,
-                            [echoDelay, echoFeedback, echoFilter]
-                        );
-                    }
-                };
+            effect: function(effect){
```

Comment:
I imagine it would be pretty difficult to learn about the specific algorithm that HyperCard used, as it involved proprietary APIs within the classic Macintosh Toolkit. I think resampling in general involves a combination of changing the sample rate (eg, if it's 44.1 kHz, treat it as 22 kHz instead) and a transformation to correct for the fact that simply changing the sample rate also changes the temporal length of the sound. If the goal is to get a specific resultant pitch, you would also need some way to determine the native pitch of the sample.